### PR TITLE
Limit concurrent creation of healthcheck gRPC connections

### PR DESCRIPTION
**Pull request by timvaillancourt**
## Description

This PR adds a concurrency limit for the number of gRPC connections created by the healthcheck handler of `vtgate`, in order to avoid hitting `runtime: program exceeds 10000-thread limit` panics

Related, in https://github.com/vitessio/vitess/pull/15030 I was able to remove 1 x sycall/stream (for gRPC client creds), however this is not enough to avoid hitting panics - now we're seeing raw network syscalls as the main thread-usage offenders

#### Hot syscall # 1 (DNS due to new gRPC connection to tablet)

```
(dlv) stack
 0  0x000000000047f120 in syscall.Syscall6
    at syscall/asm_linux_amd64.s:43
 1  0x000000000047e66c in syscall.setsockopt
    at syscall/zsyscall_linux_amd64.go:1481
 2  0x00000000005ecf33 in syscall.SetsockoptInt
    at syscall/syscall_unix.go:427
 3  0x00000000005ecf33 in net.setDefaultSockopts
    at net/sockopt_linux.go:21
 4  0x00000000005eb679 in net.socket
    at net/sock_posix.go:23
 5  0x00000000005e08f8 in net.internetSocket
    at net/ipsock_posix.go:142
 6  0x00000000005f30fd in net.(*sysDialer).dialUDP
    at net/udpsock_posix.go:206
 7  0x00000000005ca0d4 in net.(*sysDialer).dialSingle
    at net/dial.go:586
 8  0x00000000005c99b2 in net.(*sysDialer).dialSerial
    at net/dial.go:551
 9  0x00000000005c8836 in net.(*Dialer).DialContext
    at net/dial.go:428
10  0x00000000005e4485 in net.(*Resolver).dial
    at net/lookup_unix.go:70
11  0x00000000005cd905 in net.(*Resolver).exchange
    at net/dnsclient_unix.go:160
12  0x00000000005cec45 in net.(*Resolver).tryOneName
    at net/dnsclient_unix.go:260
13  0x00000000005d20c5 in net.(*Resolver).goLookupIPCNAMEOrder.func3.1
    at net/dnsclient_unix.go:612
14  0x00000000005d200a in net.(*Resolver).goLookupIPCNAMEOrder.func3.2
    at net/dnsclient_unix.go:615
15  0x00000000004692c1 in runtime.goexit
    at runtime/asm_amd64.s:1571
```

#### Hot syscall # 2 (same - gRPC connecting to tablet)

```
(dlv) stack
 0  0x000000000047f120 in syscall.Syscall6
    at syscall/asm_linux_amd64.s:43
 1  0x000000000047e40f in syscall.fstatat
    at syscall/zsyscall_linux_amd64.go:1450
 2  0x00000000004f9658 in syscall.Stat
    at syscall/syscall_linux_amd64.go:60
 3  0x00000000004f9658 in os.statNolog.func1
    at os/stat_unix.go:32
 4  0x00000000004f9658 in os.ignoringEINTR
    at os/file_posix.go:245
 5  0x00000000004f9658 in os.statNolog
    at os/stat_unix.go:31
 6  0x00000000004f9194 in os.Stat
    at os/stat.go:13
 7  0x00000000005e9425 in net.stat
    at net/parse.go:76
 8  0x00000000005d69c5 in net.readHosts
    at net/hosts.go:59
 9  0x00000000005d71a6 in net.lookupStaticHost
    at net/hosts.go:107
10  0x00000000005d0773 in net.goLookupIPFiles
    at net/dnsclient_unix.go:551
11  0x00000000005d0c54 in net.(*Resolver).goLookupIPCNAMEOrder
    at net/dnsclient_unix.go:572
12  0x00000000005e47da in net.(*Resolver).lookupIP
    at net/lookup_unix.go:102
13  0x00000000005f945b in net.(*Resolver).lookupIP-fm
    at <autogenerated>:1
14  0x00000000005d673d in net.glob..func1
    at net/hook.go:23
15  0x00000000005e2e9f in net.(*Resolver).lookupIPAddr.func1
    at net/lookup.go:319
16  0x00000000005c391b in internal/singleflight.(*Group).doCall
    at internal/singleflight/singleflight.go:95
17  0x00000000005c38b6 in internal/single flight.(*Group).DoChan.func1
    at internal/singleflight/singleflight.go:88
18  0x00000000004692c1 in runtime.goexit
    at runtime/asm_amd64.s:1571
```

# Hot syscall # 3

I lost this one unfortunately, but it was a `syscall.Open` for the client TCP connection open

## Related Issue(s)

- Issue TBD
- https://github.com/vitessio/vitess/pull/15030

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [ ] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [ ] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on CI?
-   [ ] Documentation was added or is not required

### CODE REVIEW

**Comment by timvaillancourt** [comment]
On 2nd thought, `--healthcheck-connect-concurrency` is probably a more accurate flag name. Feedback appreciated 🤔

**Comment by harshit-gangal** [comment]
should a VTGate be cell-bound then to connect to ~10k vttablets?

**Comment by timvaillancourt** [comment]
> should a VTGate be cell-bound then to connect to ~10k vttablets?

@harshit-gangal the `vtgate`s we're seeing crash do have a `--cells_to_watch`, however this is set to every potential cell the `PRIMARY` could failover to, as these pools require `PRIMARY` operations

Wherever possible our `vtgate`s subscribe to the minimum-necessary `--keyspaces_to_watch`, however on the `vtgate`s that crash, this list is either very large, or in one case not set at all for a prod-debugging use case. Perhaps with some changes to our architecture this could be improved, but not easily

File: go/vt/discovery/tablet_health_check.go [hunk 1]
```diff
@@ -136,14 +138,30 @@ func (thc *tabletHealthCheck) stream(ctx context.Context, callback func(*query.S
...
 	defer thc.connMu.Unlock()
-	return thc.connectionLocked()
+	return thc.connectionLocked(hc)
 }
 
-func (thc *tabletHealthCheck) connectionLocked() queryservice.QueryService {
+func (thc *tabletHealthCheck) connectionLocked(hc *HealthCheckImpl) queryservice.QueryService {
 	if thc.Conn == nil {
+		grpcclient.RegisterGRPCDialOptions(func(opts []grpc.DialOption) ([]grpc.DialOption, error) {
+			dialerFunc := func(ctx context.Context, addr string) (net.Conn, error) {
```
Comment (timvaillancourt) [inline go/vt/discovery/tablet_health_check.go:150]: If other parts of the code could benefit from a similar limit, perhaps this should be moved to a helper/util library or baked into `grpcclient`? 🤔 cc @deepthi (and anyone else) for thoughts

File: go/vt/discovery/tablet_health_check.go [hunk 2]
```diff
@@ -136,14 +138,32 @@ func (thc *tabletHealthCheck) stream(ctx context.Context, callback func(*query.S
...
-func (thc *tabletHealthCheck) Connection() queryservice.QueryService {
+func (thc *tabletHealthCheck) Connection(hc *HealthCheckImpl) queryservice.QueryService {
 	thc.connMu.Lock()
 	defer thc.connMu.Unlock()
-	return thc.connectionLocked()
+	return thc.connectionLocked(hc)
 }
 
-func (thc *tabletHealthCheck) connectionLocked() queryservice.QueryService {
+func healthCheckDialerFactory(hc *HealthCheckImpl) func(ctx context.Context, addr string) (net.Conn, error) {
```
Comment (timvaillancourt) [inline go/vt/discovery/tablet_health_check.go:147]: If other parts of the code could benefit from a similar limit, perhaps this should be moved to a helper/util library or baked into grpcclient? 🤔 cc @deepthi (and anyone else) for thoughts

File: go/vt/discovery/healthcheck.go [hunk 3]
```diff
@@ -168,6 +172,7 @@ func registerWebUIFlags(fs *pflag.FlagSet) {
 	fs.StringVar(&TabletURLTemplateString, "tablet_url_template", "http://{{.GetTabletHostPort}}", "Format string describing debug tablet url formatting. See getTabletDebugURL() for how to customize this.")
 	fs.DurationVar(&refreshInterval, "tablet_refresh_interval", 1*time.Minute, "Tablet refresh interval.")
 	fs.BoolVar(&refreshKnownTablets, "tablet_refresh_known_tablets", true, "Whether to reload the tablet's address/port map from topo in case they change.")
+	fs.Int64Var(&healthCheckConnectConcurrency, "healthcheck-connect-concurrency", 1024, "Concurrency of new healthcheck connections. This should be less than the golang max thread limit of 10000.")
```
Comment (timvaillancourt) [inline go/vt/discovery/healthcheck.go:175]: An alternative: `--healthcheck-dial-concurrency` 🤔

An some feedback on the `1024` default: it's working nicely for us without introducing any noticeable delays to startup or hitting the 10k thread limit

File: go/vt/grpcclient/client.go [hunk 4]
```diff
@@ -39,6 +40,7 @@ import (
 )
 
 var (
+	grpcOptionsMu         sync.Mutex
```
Comment (timvaillancourt) [inline go/vt/grpcclient/client.go:43]: Added to fix data race:
```bash
WARNING: DATA RACE
Read at 0x000003e57f70 by goroutine 71:
  vitess.io/vitess/go/vt/grpcclient.RegisterGRPCDialOptions()
      /Users/tim/github/vitess/go/vt/grpcclient/client.go:89 +0xdd
  vitess.io/vitess/go/vt/discovery.(*tabletHealthCheck).connectionLocked()
      /Users/tim/github/vitess/go/vt/discovery/tablet_health_check.go:164 +0x68
  vitess.io/vitess/go/vt/discovery.(*tabletHealthCheck).Connection()
      /Users/tim/github/vitess/go/vt/discovery/tablet_health_check.go:144 +0xb5
  vitess.io/vitess/go/vt/discovery.(*tabletHealthCheck).stream()
      /Users/tim/github/vitess/go/vt/discovery/tablet_health_check.go:128 +0x44
  vitess.io/vitess/go/vt/discovery.(*tabletHealthCheck).checkConn()
      /Users/tim/github/vitess/go/vt/discovery/tablet_health_check.go:295 +0x6ee
  vitess.io/vitess/go/vt/discovery.(*HealthCheckImpl).AddTablet.func2()
      /Users/tim/github/vitess/go/vt/discovery/healthcheck.go:426 +0x44

Previous write at 0x000003e57f70 by goroutine 69:
  vitess.io/vitess/go/vt/grpcclient.RegisterGRPCDialOptions()
      /Users/tim/github/vitess/go/vt/grpcclient/client.go:89 +0x184
  vitess.io/vitess/go/vt/discovery.(*tabletHealthCheck).connectionLocked()
      /Users/tim/github/vitess/go/vt/discovery/tablet_health_check.go:164 +0x68
  vitess.io/vitess/go/vt/discovery.(*tabletHealthCheck).Connection()
      /Users/tim/github/vitess/go/vt/discovery/tablet_health_check.go:144 +0xb5
  vitess.io/vitess/go/vt/discovery.(*tabletHealthCheck).stream()
      /Users/tim/github/vitess/go/vt/discovery/tablet_health_check.go:128 +0x44
  vitess.io/vitess/go/vt/discovery.(*tabletHealthCheck).checkConn()
      /Users/tim/github/vitess/go/vt/discovery/tablet_health_check.go:295 +0x6ee
  vitess.io/vitess/go/vt/discovery.(*HealthCheckImpl).AddTablet.func2()
      /Users/tim/github/vitess/go/vt/discovery/healthcheck.go:426 +0x44

Goroutine 71 (running) created at:
  vitess.io/vitess/go/vt/discovery.(*HealthCheckImpl).AddTablet()
      /Users/tim/github/vitess/go/vt/discovery/healthcheck.go:426 +0xad5
  vitess.io/vitess/go/vt/discovery.TestHealthCheckErrorOnPrimaryAfterExternalReparent()
      /Users/tim/github/vitess/go/vt/discovery/healthcheck_test.go:358 +0x971
  testing.tRunner()
      /usr/local/Cellar/go/1.21.7/libexec/src/testing/testing.go:1595 +0x261
  testing.(*T).Run.func1()
      /usr/local/Cellar/go/1.21.7/libexec/src/testing/testing.go:1648 +0x44

Goroutine 69 (running) created at:
  vitess.io/vitess/go/vt/discovery.(*HealthCheckImpl).AddTablet()
      /Users/tim/github/vitess/go/vt/discovery/healthcheck.go:426 +0xad5
  vitess.io/vitess/go/vt/discovery.TestHealthCheckErrorOnPrimaryAfterExternalReparent()
      /Users/tim/github/vitess/go/vt/discovery/healthcheck_test.go:351 +0x5cc
  testing.tRunner()
      /usr/local/Cellar/go/1.21.7/libexec/src/testing/testing.go:1595 +0x261
  testing.(*T).Run.func1()
      /usr/local/Cellar/go/1.21.7/libexec/src/testing/testing.go:1648 +0x44
```

File: go/vt/discovery/healthcheck.go (hunk 3 above)
Comment (timvaillancourt) [reply go/vt/discovery/healthcheck.go]: Updated to `--healthcheck-dial-concurrency`. It explains what is going on better 👍

**Comment by harshit-gangal** [comment]
This PR is slowing down the parallel vttablet connections being opened but still has the 10k limit on the thread count.

I found that `runtime/debug` has `SetMaxThreads` which can be used to modify the limit
https://pkg.go.dev/runtime/debug#SetMaxThreads

**Review by harshit-gangal — COMMENTED** [review]
Changes look fine to me. An issue and adding release notes would be good.

**Comment by timvaillancourt** [comment]
> I found that runtime/debug has SetMaxThreads which can be used to modify the limit
https://pkg.go.dev/runtime/debug#SetMaxThreads

@harshit-gangal we considered raising the max threads limit but tried to avoid that in cased it allowed a different problem to spiral out of control, also it doesn't scale as easily with growth unless it's cranked to a very high value

We've been using the `1024` default from the PR on production on gates opening 12,000+ healthcheck streams, but it's possible that we may revisit that number

Related, we've noticed a blip of `no health tablets for...` during vtgate rollout for shards that are seemingly healthy, which may or may not be patch (unable to confirm); my dive through the logic for waiting for tablets to healthcheck lead me to believe this patch wouldn't cause that issue and I'm maybe just paranoid of my recent change. If there is a race or bug in that logic, this PR could make matters worse, however. But any 💡s what might cause this is appreciated 🙇

**Comment by timvaillancourt** [comment]
@harshit-gangal / @vmg I believe the doc changes are ready 🙇

**Comment by vmg** [comment]
Thanks Tim! LGTM. I'm 99% sure the flags section in the website is auto-generated now. @frouioui will know for sure.

Need +1 to merge @harshit-gangal

**Review by harshit-gangal — APPROVED** [review]

**Comment by harshit-gangal** [comment]
@timvaillancourt now it needs a conflict resolution.

**Pull request merged**